File: 07_3_ants_neural_net.py
# Ant data: neural network
# Brett Melbourne
# 27 Feb 2024

# A single layer neural network, or feedforward network, illustrated with the
# ants data. We first hand code the model algorithm as a proof of understanding.
# Then we code the same model and train it using Keras.

import numpy as np
import pandas as pd
from plotnine import *
from source.ml4e import expand_grid

# Ant data with 3 predictors of species richness
ants = pd.read_csv("data/ants.csv")
ants = ants[["richness", "latitude", "habitat", "elevation"]]
ants.head()

# Scaling parameters
lat_mn = np.mean(ants["latitude"])
lat_sd = np.std(ants["latitude"], ddof=1) #sample sd, same as R code
ele_mn = np.mean(ants["elevation"])
ele_sd = np.std(ants["elevation"], ddof=1)


# Hand-coded feedforward network

# Before we go on to Keras, we're going to hand code a feedforward network in
# Python to gain a good understanding of the model algorithm. Here is our
# pseudocode from the lecture notes.

# Single layer neural network, model algorithm:
#
# define g(z)
# load X, i=1...n, j=1...p (in appropriate form)
# set K
# set weights and biases: w(1)_jk, b(1)_k, w(2)_k1, b(2)_1
# for each activation unit k in 1:K
#     calculate linear predictor: z_k = b(1)_k + Xw(1)_k
#     calculate nonlinear activation: A_k = g(z_k)
# calculate linear model: f(X) = b(2)_1 + Aw(2)_1
# return f(X)


# Now code this algorithm in Python. I already trained this model using Keras
# (see later) to obtain a parameter set for the weights and biases.

# Single layer neural network, model algorithm

# define g(z)
def g_relu(z):
    g_z = np.where(z < 0, 0, z)
    return g_z

# load x (could be a grid of new predictor values or the original data)
grid_data = expand_grid(
    {"latitude": np.linspace(np.min(ants["latitude"]), np.max(ants["latitude"]), 201),
     "habitat":  ["forest","bog"],
     "elevation": np.linspace(np.min(ants["elevation"]), np.max(ants["elevation"]), 51)}
)

# data preparation: scale, dummy encoding, convert to matrix
x = grid_data.copy()
x["latitude"] = (x["latitude"] - lat_mn) / lat_sd
x["elevation"] = (x["elevation"] - ele_mn) / ele_sd
x["bog"] = np.where(x["habitat"] == "bog", 1, 0)
x["forest"] = np.where(x["habitat"] == "forest", 1, 0)
x = x[["latitude", "bog", "forest", "elevation"]]
x = x.to_numpy()

# dimensions of x
n = x.shape[0] #rows
p = x.shape[1]

# set K
K = 5

# set parameters (weights and biases)
w1 = np.array([[-0.2514450848, 0.4609818,  0.1607399, -0.9136779, -1.0339828],
               [-0.4243144095, 0.7681985, -0.1529205, -0.3439012,  0.8026423],
               [-0.0005548226, 0.6407318,  1.4387618,  1.6372939,  1.1695395],
               [-0.0395327508, 0.5222837, -0.6239772, -0.3365386, -0.7156096]])

b1 = np.array([-0.2956778, 0.3149067, 0.8800480, 0.6910487, 0.6947369])
b1 = [-0.2956778, 0.3149067, 0.8800480, 0.6910487, 0.6947369]

w2 = np.array([[-0.4076283],
               [ 0.6379358],
               [ 0.8768858],
               [ 1.6320601],
               [ 0.9864114]])

b2 = 0.8487959


# hidden layer 1, iterating over each activation unit
A = np.full((n,K), np.nan)
for k in range(K):
#   linear predictor (dot is matrix multiplication in numpy)
    z = x.dot(w1[:,k]) + b1[k]
#   nonlinear activation
    A[:,k] = g_relu(z)

# output, layer 2, linear model
f_x = A.dot(w2) + b2

# return f(x)
f_x


# Plot predictions
nn1_preds = pd.DataFrame(f_x, columns=["richness"])
preds = pd.concat([grid_data, nn1_preds], axis=1)

(ggplot()
+ geom_line(aes(x="latitude", y="richness", color="elevation",
                group="factor(elevation)"),
                data=preds, linetype="dashed")
+ geom_point(aes(x="latitude", y="richness", color="elevation"),
                data=ants)
+ facet_wrap("~habitat")
+ theme_bw())


# Using Keras to fit neural networks

# You'll need to install Python and Tensorflow, which will also install the
# keras Python library. The keras library is an interface to the Python
# Tensorflow library, which in turn is an interface to Tensorflow (mostly C++)!
# See Assignment 4 for installation directions into a conda environment.

import tensorflow as tf
import keras
from keras import layers
import matplotlib.pyplot as plt

# Set random seed for reproducibility
keras.utils.set_random_seed(5574)

# Prepare data
xtrain = ants.copy()
xtrain["latitude"] = (xtrain["latitude"] - lat_mn) / lat_sd
xtrain["elevation"] = (xtrain["elevation"] - ele_mn) / ele_sd
xtrain["bog"] = np.where(xtrain["habitat"] == "bog", 1, 0)
xtrain["forest"] = np.where(xtrain["habitat"] == "forest", 1, 0)
xtrain = xtrain[["latitude", "bog", "forest", "elevation"]]
xtrain = xtrain.to_numpy()

ytrain = ants["richness"]

# Specify the model
modnn1 = keras.Sequential()
modnn1.add(keras.Input(shape=xtrain.shape[1]))
modnn1.add(layers.Dense(units=5))        
modnn1.add(layers.Activation("relu"))
modnn1.add(layers.Dense(units=1))

# Check configuration
modnn1.summary()

# Compile and train the model
modnn1.compile(optimizer="rmsprop", loss="mse")
history = modnn1.fit(xtrain, ytrain, epochs=300, batch_size=4)
history = pd.DataFrame(history.history)

# To visualize model fit in real time, you can use tensorboard
# https://keras.io/api/callbacks/tensorboard/
# https://www.tensorflow.org/tensorboard/get_started

# Save model and history, or load it back in
# tf.saved_model.save(modnn1, "07_3_ants_neural_net_files/saved/modnn1")
# history.to_csv("07_3_ants_neural_net_files/saved/modnn1_history.csv", index=False)
# modnn1 = keras.models.load_model("07_3_ants_neural_net_files/saved/modnn1")
# history = pd.read_csv("07_3_ants_neural_net_files/saved/modnn1_history.csv")

# Plot history
history.plot()
plt.show()

# Make predictions for predictor grid
npred = modnn1.predict(x)
npred = pd.DataFrame(npred, columns=["richness"])
preds = pd.concat([grid_data, npred], axis=1)

(ggplot()
+ geom_line(aes(x="latitude", y="richness", color="elevation",
                group="factor(elevation)"),
                data=preds, linetype="dashed")
+ geom_point(aes(x="latitude", y="richness", color="elevation"),
                data=ants)
+ facet_wrap("~habitat")
+ theme_bw())    

# Get weights and biases
modnn1.get_weights()

File: 07_3_ants_neural_net.R
#' ---
#' title: "Ant data: neural network"
#' author: Brett Melbourne
#' date: 27 Feb 2024
#' output:
#'     github_document
#' ---

#' A single layer neural network, or feedforward network, illustrated with the
#' ants data. We first hand code the model algorithm as a proof of
#' understanding. Then we code the same model and train it using Keras.
#' 

#+ results=FALSE, message=FALSE, warning=FALSE
library(ggplot2)
library(dplyr)

#' Ant data with 3 predictors of species richness

ants <- read.csv("data/ants.csv") |> 
    select(richness, latitude, habitat, elevation)
head(ants)

#' For neural networks it is customary to scale numeric predictors to zero mean,
#' unit variance, as the training algorithms perform better. To make predictions
#' with new data, we need to use the scaling parameters from the data used to
#' train the model. Here, we calculate the scaling parameters (mean and standard
#' deviation) to use later.

lat_mn <- mean(ants$latitude)
lat_sd <- sd(ants$latitude)
ele_mn <- mean(ants$elevation)
ele_sd <- sd(ants$elevation)


#' ### Hand-coded feedforward network

#' Before we go on to Keras, we're going to hand code a feedforward network in R
#' to gain a good understanding of the model algorithm. Here is our pseudocode
#' from the lecture notes.
#' 

#' ```
#' Single layer neural network, model algorithm:
#'
#' define g(z)
#' load X, i=1...n, j=1...p (in appropriate form)
#' set K
#' set weights and biases: w(1)_jk, b(1)_k, w(2)_k1, b(2)_1
#' for each activation unit k in 1:K
#'     calculate linear predictor: z_k = b(1)_k + Xw(1)_k
#'     calculate nonlinear activation: A_k = g(z_k)
#' calculate linear model: f(X) = b(2)_1 + Aw(2)_1
#' return f(X)
#' ```
#' 

#' Now code this algorithm in R. I already trained this model using Keras (see
#' later) to obtain a parameter set for the weights and biases.

# Single layer neural network, model algorithm

# define g(z)
g_relu <- function(z) {
    g_z <- ifelse(z < 0, 0, z)
    return(g_z)
}

# load x (could be a grid of new predictor values or the original data)
grid_data  <- expand.grid(
    latitude=seq(min(ants$latitude), max(ants$latitude), length.out=201),
    habitat=c("forest","bog"),
    elevation=seq(min(ants$elevation), max(ants$elevation), length.out=51))

# data preparation: scale, one-hot encoding, convert to matrix
x <- grid_data |>
    mutate(latitude = (latitude - lat_mn) / lat_sd,
           elevation = (elevation - ele_mn) / ele_sd,
           bog = ifelse(habitat == "bog", 1, 0),
           forest = ifelse(habitat == "forest", 1, 0)) |>    
    select(latitude, bog, forest, elevation) |>     #drop richness & habitat
    as.matrix()

# dimensions of x
n <- nrow(x)
p <- ncol(x)

# set K
K <- 5

# set parameters (weights and biases)
w1 <- c(-0.2514450848, 0.4609818,  0.1607399, -0.9136779, -1.0339828,
        -0.4243144095, 0.7681985, -0.1529205, -0.3439012,  0.8026423,
        -0.0005548226, 0.6407318,  1.4387618,  1.6372939,  1.1695395,
        -0.0395327508, 0.5222837, -0.6239772, -0.3365386, -0.7156096) |>
    matrix(nrow=4, ncol=5, byrow=TRUE)

b1 <- c(-0.2956778, 0.3149067, 0.8800480, 0.6910487, 0.6947369)

w2 <- c(-0.4076283,
         0.6379358,
         0.8768858,
         1.6320601,
         0.9864114) |> 
    matrix(nrow=5, ncol=1, byrow=TRUE)

b2 <- 0.8487959


# hidden layer 1, iterating over each activation unit
A <- matrix(NA, nrow=n, ncol=K)
for ( k in 1:K ) {
#   linear predictor
    z <- x %*% w1[,k] + b1[k]
#   nonlinear activation
    A[,k] <- g_relu(z)
}

# output, layer 2, linear model
f_x <- A %*% w2  + b2

# return f(x); a redundant copy but mirrors our previous examples
nn1_preds <- f_x


#' Plot predictions

preds <- cbind(grid_data, richness=nn1_preds)
ants |> 
    ggplot() +
    geom_line(data=preds, 
              aes(x=latitude, y=richness, col=elevation, group=factor(elevation)),
              linetype=2) +
    geom_point(aes(x=latitude, y=richness, col=elevation)) +
    facet_wrap(vars(habitat)) +
    scale_color_viridis_c() +
    theme_bw()


#' ### Using Keras to fit neural networks

#' Now we'll consider the very same model with Keras, including training the
#' model on the data to determine the weight and bias parameters.
#' 

#' You'll need to install Python and Tensorflow to use the `keras` package. The
#' `keras` package is an R interface to the Python Keras library, which in turn
#' is an interface to the Python Tensorflow library, which in turn is an
#' interface to Tensorflow (mostly C++)! See Assignment 4 for installation
#' directions. The Python Keras library is widely used and the R functions and
#' workflow closely mirror the Python functions and workflow, so what we'll
#' learn in Keras for R largely applies to Keras for Python as well. See also
#' the Python version of this script.
#' 

#+ warning=FALSE
# If you have another conda environment as your R default, you will need the
# following line before loading the keras library
reticulate::use_condaenv(condaenv = "r-tensorflow")
library(keras)

#' First we'll set a random seed for reproducibility. The seed applies to R,
#' Python, and Tensorflow. It will take a few moments for Tensorflow to get set
#' up.

tensorflow::set_random_seed(5574)

#' Next, prepare the data (the predictors are prepared exactly as we did for the
#' hand-coded version above):
 
xtrain <- ants |> 
    mutate(latitude = (latitude - lat_mn) / lat_sd,
           elevation = (elevation - ele_mn) / ele_sd,
           bog = ifelse(habitat == "bog", 1, 0),
           forest = ifelse(habitat == "forest", 1, 0)) |>    
    select(latitude, bog, forest, elevation) |>     #drop richness & habitat
    as.matrix()

ytrain <- ants[,"richness"]

#' Next, specify the model. The basic syntax builds the model layer by layer,
#' using the pipe operator to express the flow of information from the output of
#' one layer to the input of the next. Here, `keras_model_sequential()` says we
#' will build a sequential (or feedforward) network and the input layer will
#' have `ncol(xtrain)` units (i.e. there will be one input unit, or node, for
#' each of the 4 predictor columns in our ant data). The next layer, a hidden
#' layer, will be a densely-connected layer (i.e. all units from the previous
#' layer connected to all units of the hidden layer) with 5 units, and it will
#' be passed through the ReLU activation function. The output layer will be
#' another densely-connected layer with 1 unit and no activation applied.

modnn1 <- keras_model_sequential(input_shape = ncol(xtrain)) |>
    layer_dense(units = 5) |>
    layer_activation("relu") |> 
    layer_dense(units = 1)

#' We can check the configuration:

modnn1

#' The model and layer names, such as "dense_1" and "activation" are names
#' given to the objects created in the tensorflow workspace for this R session.
#' We see that layer 1 has 25 parameters as we expect (5x4 weights and 5 biases)
#' and the output layer has 6 parameters as we expect (5 weights and 1 bias).
#' This model thus has a total of 31 parameters.
#' 

#' Next, compile the model. We are specifying that the optimizer algorithm is
#' RMSprop and the loss function is mean squared error. Notice that we are not
#' modifying the R `modnn1` object but are instead directing Keras in Python to
#' set up for training the model.

compile(modnn1, optimizer="rmsprop", loss="mse")

#' The RMSprop algorithm is the default in Keras and works for most models. It
#' implements stochastic gradient descent with an adaptive learning rate and
#' various performance enhancements. By default, the initial learning rate is
#' 0.001 (see `?optimizer_rmsprop`) but this can be tuned. Other optimizers are
#' available.
#'

#' Now train the model, keeping a copy of the training history. Again, we are
#' not getting a new R fitted-model object out of this as we would in, say, a
#' call to `fit()` with an `lm` object but instead we are directing Keras in
#' Python to train the model (all the important output remains in Python data
#' structures). The history object is a by-product of training, so I put it on
#' the right-hand side of the expression to be clear that this is a by-product
#' and not a traditional R fitted-model object (where you would usually expect
#' to find the trained parameter values). In one epoch, the training data are
#' sampled in batches (one SGD step is taken for each batch) until all of the
#' data have been sampled, so the number of epochs is the number of complete
#' iterations through the training data. I chose 300 epochs because the fit
#' improves only slowly beyond that. Here I used a batch size of 4, which means
#' 10% of the data are used on each subsample to calculate the stochastic
#' gradient descent step. Training will take a minute or so and a plot will
#' chart its progress.

#+ eval=FALSE

fit(modnn1, xtrain, ytrain, epochs = 300, batch_size=4) -> history

#' As it takes time to train these models, it's worth saving the model
#' (Tensorflow format) and history (R format) so they they can be reloaded
#' later. We can also load this saved model in Python or share with colleagues.

# save_model_tf(modnn1, "07_3_ants_neural_net_files/saved/modnn1")
# save(history, file="07_3_ants_neural_net_files/saved/modnn1_history.Rdata")
modnn1 <- load_model_tf("07_3_ants_neural_net_files/saved/modnn1")
load("07_3_ants_neural_net_files/saved/modnn1_history.Rdata")

#' We can plot the history once training is done. If you have `ggplot2` loaded,
#' it will create a ggplot, otherwise it will create a base plot.

plot(history, smooth=FALSE, theme_bw=TRUE)

#' We want to see the training error (RMS) decline to a reasonable level.
#' Although the error will continue to go down, here we see it leveling out
#' somewhat at about RMS=7, which is an absolute error of sqrt(7) = +/- 2.6
#' species.
#' 

#' Make predictions for our grid of new predictor variables (`x`; we made this
#' earlier, scaling by the mean and sd of the data) and plot the fitted model
#' with the data. This plot is exactly the same as the one we produced earlier
#' "by hand" since they are the same model.

npred <- predict(modnn1, x)
preds <- cbind(grid_data, richness=npred)
ants |> 
    ggplot() +
    geom_line(data=preds, 
              aes(x=latitude, y=richness, col=elevation,
                  group=factor(elevation)),
              linetype=2) +
    geom_point(aes(x=latitude, y=richness, col=elevation)) +
    facet_wrap(vars(habitat)) +
    scale_color_viridis_c() +
    theme_bw()

#' We can get the weights and biases, returned as a list

get_weights(modnn1)
